---
title: "Example code for single-cell analysis with Seurat, day 1"
author: "Kevin Rue-Albrecht"
date: "05/10/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Seurat)
library(tidyverse)

```

# Exercise

## Import scRNA-seq data and create a Seurat object

- Load the `Seurat` package.

```{r}
read10x_data <- Read10X("/project/obds/shared/resources/4_r_single_cell/singlecell_seuratday1/filtered_feature_bc_matrix")
```

- Use the function `Read10X()` to import data in the directory `filtered_feature_bc_matrix/` 
  as an object named `read10x_data`.
  What class of object does the function return?

```{r}
class(read10x_data) #the same gene are represented multiple times as they can be found on different chromosomes. You can change it to match to Ensembl ID by changing gene.column value, but it may be better do this later on once you have found out what's interesting (e.g if you wanted to do a pathway analysis).
```

> Answer:
> mgCMatrix - it is not a seurat object initially

- Have a look at the object and its structure (e.g., first 15 rows and 6 columns).
  What is a sparse matrix and how does it store data?

```{r}
dim(read10x_data)
```

```{r}
read10x_data[1:15,1:6] 
```

> Answer:
> #all the "." represent zeros - this is a sparse data set. You need to store 3 values - which row, which column, and which value. By only including this information for non-zero values it saves memory space as there is so many zeros.

- How many features and barcodes (i.e., cells) are present in the data set?

```{r}
dim(read10x_data)
```

> Answer:
> 33538 features (genes) and 5155 barcodes (cells_

- Create a Seurat object using the function `CreateSeuratObject()`
  and the object `read10x_data`.
  Name the object `seurat_object`.
  Include features detected in at least 3 cells,
  and cells where at least 200 features detected.
  Name the project `pbmc5k`.
  How many features and barcodes are left in the Seurat object?

```{r}
seurat_object <- CreateSeuratObject(
    counts = read10x_data,
    project = "pbmc5k", #this is an arbirary identifier that you name.project is one data set, it is only really useful if you need to merge datasets.
    min.cells = 3, #these are arbitrary values that we have selected, there may be a default Seurat uses
    min.features = 200 #these are arbitrary values that we have selected, there may be a default Seurat uses
)
seurat_object
```

> Answer:
> 19037 features across 5100 samples (barcodes) within 1 assay (an assay is just one stage of processing the object) 

- How many features and cells were filtered out when you created the Seurat object?

```{r}
dim(read10x_data) - dim(seurat_object) #you can subtract matrixes to figure out how many features and cells you have lost. You say barcodes rather than cells as there can be doublets (two cells) that have the same barcode.
```

> 

# Exercise

## Accessing the contents of a Seurat object

- Query the name of the default assay in the Seurat object.

```{r}
DefaultAssay(seurat_object)
```

- List the names of assays available in the Seurat object.

```{r}
Assays(seurat_object)
str(seurat_object$RNA)
```

- Display the first six rows and six columns of the RNA assay data.
  What function do you use?
  Which arguments do you need to supply, and which ones are optional?

```{r}
seurat_object[[]]

GetAssayData(seurat_object, assay = "RNA", slot = "data")[1:6,1:6]
```

> Answer:
> 

- Display the entire data.frame of per-cell metadata (first six rows).
  What column names do you see?

```{r}
seurat_object[[]][1:6,]
```

> Answer:
> 

- Fetch one column of metadata using `[[`.
  What type of object do you get back?

```{r}
class(seurat_object[["nCount_RNA"]])

```

> Answer:
> 

- Instead,fetch the same column of metadata using `$`.
  What type of object do you get back this time?

```{r}
class(seurat_object$nCount_RNA)
```

> Answer:
> 

- Use the function `FetchData()` to access the library size and expression of the feature named `"LYZ"` and `"nCount_RNA"' (first six rows).
  What type of object do you get back?

```{r}
FetchData(object = seurat_object, vars = c("LYZ", "nCount_RNA", slot = "data"))[1:6,] #this gives a datagrame which means you can use it in ggplot.
```

> 

# Demo

## Common operations on Seurat objects

`WhichCells()` returns the names of cells that match a logical expression.

```{r}
WhichCells(seurat_object, expression = LYZ > 500)
```

`VariableFeatures()` returns the names of variable features (for a given assay, if computed).

```{r}
VariableFeatures(seurat_object)
```

`subset()` returns a new Seurat object restricted to certain features and cells.

```{r}
subset(
    x = seurat_object,
    cells = WhichCells(seurat_object, expression = LYZ > 500),
    features = VariableFeatures(object = seurat_object)
)
```

# Exercise

## Quality control and visualisation

- The library size and number of features detected per cell is already present in the Seurat object.
  Use the function `VlnPlot()` to display them in a single violin plot.

```{r}
VlnPlot(object = seurat_object, features = "nFeature_RNA")
```

- Use the function `PercentageFeatureSet()` to compute the fraction of reads
  assigned to mitochondrial genes in each cell.
  Store the metric in the cell metadata of the Seurat object, under the name `"percent_mt"`.
  Visualise this new metric alongside the previous two in a new violin plot.

```{r}
seurat_object[["percent_mt"]] <- PercentageFeatureSet(seurat_object, pattern = "^MT-")

seurat_object[[]] #then check that it's not messed up the other columns

VlnPlot(object = seurat_object, features = "percent_mt")
```

- Visualise a scatter plot of the proportion of mitochondrial UMIs against the library size in each cell.

```{r}
FeatureScatter(seurat_object, feature1 = "percent_mt", feature2 = "nCount_RNA") #nCount_RNA is the number of bits of RNA that make up the library.
```

- Create a new Seurat object, called `seurat_after_qc`, that is subsetted to cells that have more than 4,500 UMI counts, less than 15% of UMI counts assigned to mitochondrial features, and more than 1,500 features detected.
  How many cells were removed in this step?

```{r}
seurat_after_qc <- subset(
    seurat_object,
    subset = nCount_RNA > 4500 & percent_mt < 15 & nFeature_RNA > 1500
)
seurat_after_qc

seurat_after_qc_not_norm <- seurat_after_qc #nb this is for later
```

```{r}
dim(seurat_object) - dim(seurat_after_qc)
```

> Answer:
> 

# Exercise

## Normalisation

- Normalise the RNA assay of the Seurat object (after quality control) using the `"LogNormalize"` method.

```{r}
seurat_after_qc <- NormalizeData(seurat_after_qc,
                                 normalization.method = "LogNormalize")
                                 
head(seurat_after_qc)                                 
  # the first time you do something use x, check you are happy with it and then replace with your seurat object.
  
```

## Bonus

- Visualise the distribution of raw counts and normalised data for a feature of your choice.

```{r}
FetchData(seurat_after_qc, vars= c("nCount_RNA", "LYZ"), slot = "data")

ggplot_lyz_raw <- ggplot(data = FetchData(seurat_after_qc_not_norm, vars= c("nCount_RNA", "LYZ"), slot = "data"),
                         aes(x=LYZ)
                         ) +
    geom_histogram(   ) +
    coord_cartesian(ylim = c(0, 500)) +
    cowplot::theme_cowplot()
ggplot_lyz_normalised <- ggplot(data = FetchData(seurat_after_qc, vars= c("nCount_RNA", "LYZ"), slot = "data"),
                         aes(x=LYZ)
                         ) +
    geom_histogram(   ) +
    coord_cartesian(ylim = c(0, 500)) +
    cowplot::theme_cowplot()
cowplot::plot_grid(ggplot_lyz_raw, ggplot_lyz_normalised, ncol = 1)

#you can use this to check that each gene was normalised properly. If youdon't normalise thinks like cytokines are much more variable than a transcription factor.
```

# Exercise

## Variable features and scaling

- Identify variable features in the normalised RNA assay of the Seurat object.
  Use the `"vst"` method and select the 2,000 most variable features.
  What does this subsetting do, and what are our motivations for doing it?

```{r}
#the only constant genes are those that aren't expressed
seurat_after_qc <- FindVariableFeatures(object = seurat_after_qc,
                                        selection.method = "vst",
                                        nfeatures = 2000
) #the number of variable genes might mean you lose cell populations in the future if you haven't selected enough variable genes for that cell population.
#2000 is an arbitrary value that is selected to cut the genes that are most variable i.e interesting, you can adapt it based on looking at the VariableFeaturesPlot.
```

> Answer:
> 

- What is the function to display the name of variable features in a Seurat object (e.g., first 10)?
  How can you control which assay the variable features are pull from?

```{r}
VariableFeatures(seurat_after_qc)
```

> Answer:
> 

- Use the function `VariableFeaturePlot()` to visualise the scatter plot of standardised variance against average expression.
  How would you use this plot?

```{r}
VariableFeaturePlot(seurat_after_qc) #in this figure you want to capture most of the cells above the flat line. It is possibly a bit over filtered but that doesn't matter too much it can just introduce a little bit of noise into the PCA later.

```

> Answer:
> 

- Scale the normalised RNA assay of the Seurat object, regressing the library size and the fraction of UMI counts assigned to mitochondrial features.
  What are the motivations for removing those two sources of variation?

```{r}
seurat_after_qc <- ScaleData(seurat_after_qc,
                             vars.to.regress = c("nCount_RNA","percent_mt")
)

```

> Answer:
> 

# Exercise

## Dimensionality reduction

- Run a principal component analysis on the Seurat object.
  Which features are used by the method in the default settings?
  How could you change this?
  How do you read the message output of the function `RunPCA()`?

```{r}
seurat_after_qc <- RunPCA(seurat_after_qc)
```

> Answer:
> 

- List the names of dimensionality reduction results available in the Seurat object.

```{r}
Reductions(seurat_after_qc)
```

- Use `PCAPlot()` or `DimPlot()` to produce a scatterplot of the first and second PCA components.

```{r}
PCAPlot(seurat_after_qc,
        dims = c(1,2)
        )
```

## Bonus

- Make a scatterplot of the first and second PCA components yourself using `ggplot2`.

```{r}
# Use this code chunk to prepare a data.frame for ggplot2
#pca_data <- FetchData(seurat_after_qc, vars = c(, ), slot = "data"))
#head(pca_data)
```

```{r}
ggplot(pca_data,   ) +
    geom_point(size = 0.2) +
    cowplot::theme_cowplot()
```

- Visualise the amount of variance explained the top principal components (number of your choice).
  How many principal components would you use for downstream analyses?

```{r}
ElbowPlot(seurat_after_qc, ndims = 50)
```

> 

- Run the UMAP technique on your selected number of principal components and visualise the result as a scatterplot.

```{r}
seurat_after_qc <- RunUMAP(seurat_after_qc, reduction = "pca", dims = 1:20) #in this one you have to specify 1:20 rather than 20 which is principle component 20.
UMAPPlot(seurat_after_qc)

#this can change, so it is important to save RDS the object.
saveRDS(seurat_after_qc, file = "Seuratobject_after_qc_2023_02_20.rds")
```

# Exercise

## Clustering

- Compute the graph of nearest neighbours using the function `FindNeighbors()`.
  Which principal components are used by default?
  Instead, specify the number of principal components that you have chosen earlier.

```{r}
seurat_after_qc <- FindNeighbors(object = seurat_after_qc,
                                 reduction = "pca", #this is a default for 'Seurat'
                                 dims = 1:20 #default is 1:10, we used 1:20 earlier
                                 )
```

> Answer:
> 
> The help page states that the function `FindNeighbors()` uses principal components 1 through 10, by default.

- What are the names of the nearest neighbour graphs that are now stored in the Seurat object?

```{r}
Graphs(seurat_after_qc)
```

- Finally, compute cluster labels.
  What is the default setting for the `resolution` argument?
  Instead, set it to `0.5`.
  Do you expect more or fewer clusters following that change?
  What other parameters would you also try to experiment with?

```{r}
seurat_after_qc <- FindClusters(seurat_after_qc,
                                graph.name = "RNA_snn",
                                resolution=c(0.5,0.8) #can put mutliple things in this vector
                                )

Graphs(seurat_after_qc)
```

> 

- Visualise the cluster labels on the UMAP scatter plot.
  How would you describe the agreement between the UMAP layout and the clustering results?

```{r}
UMAPPlot(
  seurat_after_qc,
  group.by = "RNA_snn_res.0.5",
  label = TRUE
)
```

> 

# Exercise

## Identify cluster markers

- Use the function `FindAllMarkers()` to identify
  positive markers for all clusters,
  filtering markers that are detected in at least 25% of the cluster,
  and with a log fold-change greater than `0.25`.
  Assign the result to an object named `seurat_markers_all`.
  What is the class of that object?
  How do you control the set of clusters that are used?

```{r}
Idents(seurat_after_qc) <- "RNA_ssn_res.0.5" #this is because we have two resolutions and two idents

seurat_markers_all <- FindAllMarkers(seurat_after_qc,
                                     logfc.threshold = 0.25,
                                     min.pct = 0.25
)
class(seurat_markers_all)
```

> Answer:
> 

- How do you read the contents of the object `seurat_markers_all`?
  How do you know which features are the markers of each cluster?

```{r}
head(seurat_markers_all)
```

> Answer:
> 

- Filter and display the top 10 markers for cluster 3.

```{r}
library(tidyverse)

seurat_markers_top_10 <- seurat_markers_all %>% filter(cluster == 3) %>% 
    arrange(desc(avg_log2FC)) %>% #can use abs as we have -ve values - but its not useful looking at negative genes sometime.
    head(10) #use head instead of top_n as top_n does some sorting in the background and you have to specificy which column you want.
seurat_markers_top_10

seurat_markers_top_10 <- seurat_markers_top_10 %>% pull(gene) #nb you can put this in the command above it just selects the best genes.
seurat_markers_top_10
  
  #ignore the rowname, there is a separate gene that is listed
```

- Visualise the expression of the top 4 marker for cluster 3 on a UMAP layout.

```{r}

FeaturePlot(seurat_after_qc,
            features = seurat_markers_top_10[1:4],
            label = TRUE
)
```

- Visualise the expression of those same 4 marker as a violin plot.
  Do you have any particular preference between the two types of plots?

```{r}
VlnPlot(seurat_after_qc,
            features = seurat_markers_top_10[1:4]
)
```

> Answer:
> 

- Use `DoHeatmap()` to visualise the top 10 (positive) markers for each cluster.
  Hint: you may want to use the function `dplyr::group_by()`.

```{r}
markers_top10_clusters <- seurat_markers_all %>%
    group_by(cluster) %>% 
    filter(p_val_adj <0.05) %>% 
    filter(avg_log2FC > 0) %>%
    arrange(desc(avg_log2FC)) %>%
    head(10) %>%
    pull(gene)
    
DoHeatmap(seurat_after_qc,
          features = markers_top10_clusters
)
```
