---
title: "DESeq2 Tutorial"
author: "Charlie George & Modified by Devika Agarwal"
date: "`r Sys.Date()`"
output: html_document
---

Useful resources 

- **DESeq2 Vignette** - http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
- **Bioconductor RNAseq Workflow Book** - http://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html


# install packages 

- If i didn't already have these installed in my environment - I would delete the # on lines 19-23 and 26-39 and run the code 

```{r}
# from cran 
# install.packages('tidyverse')
# install.packages("RColorBrewer")
# install.packages('pheatmap')
# if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

# from bioconductor
# BiocManager::install("ashr")
# BiocManager::install("apeglm")
# BiocManager::install("DESeq2")
# BiocManager::install("org.Mm.eg.db")
```

# load packages

- what do these packages do? Add comments to remind you what they do 
```{r}
library("pheatmap")   
library("RColorBrewer") #this is for making graphs looking nice,e.g. gradients etc - good for heatmaps
library(tidyverse)     
library(DESeq2)      
library(ashr)          
library(apeglm)      
library(org.Mm.eg.db)
```

# 1 Prepping our input data to make deseq2dataset object

## Import counts table

- We made this (or a very similar version) using featurecounts in week1 on the cluster
- raw data -> fastqc -> hisat2 -> featurecounts 

```{r}
# Import counts data to variable called featurecounts_table
# The file has a comment line we need to ignore at top of file 
featurecounts_table <- read_tsv("/t1-data/project/obds/shared/resources/3_r_stats_genomics/rnaseq_differential_expression/allsamples_counts.tsv.gz", comment = "#")


# lets save the gene info (Chr,Start,End,Strand, Length) columns as a seperate dataframe (df) incase we need it later
# call this table gene_info and set row names as Geneid
gene_info <- featurecounts_table %>%
                column_to_rownames("Geneid") %>%                     
                dplyr::select(Chr, Start, End, Strand, Length)


# create a counts table without the gene info (Chr,Start,End,Strand, Length) columns
# to get a clean counts table where:
# rows = genename, columns = samples
str(featurecounts_table)
counts_table <- featurecounts_table %>%
                  column_to_rownames("Geneid") %>%
                  dplyr::select(starts_with("ERR")) %>% #ERR is what are the column names start with, could also use-c(Chr,Start,End,Strand,Length), or ends_with .bam (but we will be removing .bam later)
                  as.matrix()

counts_table

# check the dimensions of this counts_table table using dim() -> should have 39,179 genes and 8 cols 

dim(counts_table) #this step is important as by using starts_with or ends_with you could miss some files if they are typos in the names.

```

# Right now lets import and set up our table of sample information

```{r}
sample_info <- read_tsv("/t1-data/project/obds/shared/resources/3_r_stats_genomics/rnaseq_differential_expression/PRJEB18572_sample_info.tsv") %>% 
  column_to_rownames("sample_id")
sample_info

# What class is the cell_type currently? 

class(sample_info$cell_type) #this is neater than manually checking str(sample_info)

# We want to change it to a factor with 2 levels (the hypothesis we will test)
# Which should be our reference level? 

unique(sample_info$cell_type) #this gives CD4 and CD8
sample_info$cell_type <- factor(sample_info$cell_type, levels = c("CD4", "CD8")) #labels can be used to collapse multiple levels that mean the same thing into 1 label, e.g. male, Male, man, M --> into one label of male.
#CD4 is the reference level as it is first, if you don't specify it will take it alphabetically.
#as.factor doesn't have as many options as factor, factor builds a new factor from things you give it.

# what other columns might we want to include in our design that we want to change to factors? 

```

# Check the sample names are the same and in the same order in the sample_info and counts_table 

```{r}
# Column names of the count table and row names of the sample info table must be in the same order
# Check this 

identical(rownames(sample_info), colnames(counts_table))

rownames(sample_info)
colnames(counts_table)

# make sample_info row names from sample_id 

# remove .bam from colnames 
colnames(counts_table) <- gsub(".bam","", colnames(counts_table)) #gsub is a function to remove suffixes and prefixes, it takes the thing you want to replace, what you want to replace, and then the data. You need to use colnames on the object before the <- sign and on the data in the table, otherwise it will replace it in the whole table.
identical(rownames(sample_info), colnames(counts_table))

rownames(sample_info)
colnames(counts_table)
```

# Generate DESeqDataSet object

```{r}
# look up function to use from slides, find help and create an object called dds
dds <- DESeqDataSetFromMatrix(countData = counts_table,
                              colData = sample_info,
                              design = ~cell_type + genotype )
  
```

## Investigate the dds object 

```{r}
# what does it look like? - e.g. use str(dds)
str(dds)

# how to access counts
assays(dds)$counts #also counts(dds) works

# access colData i.e the sample info table
colData(dds)

# access rowData
rowData(dds) #we have not added in the gene info so this is why there is no rowdata

# access the design
design(dds)

# does it have any rowRanges()?
rowRanges(dds)

# what are its rownames and colnames?
rownames(dds)
colnames(dds)
# what attributes does it have?
attributes(dds)

```

## How to subset a DESeqDataSet Object 

```{r}
# practice subsetting dds object
# select first 3 columns
sub_dds <- dds[,1:3]
```

# Filter out unexpressed genes 

- It's good practice to filter out genes with 0 counts across all samples at the start 
- This reduces the size of the object and makes computations quicker without affecting the results as these genes are filtered out anyway in the analysis 
- Note that more strict filtering to increase power is automatically applied via independent filtering on the mean of normalized counts within the results function.


```{r}
# filter out genes that have 0 counts across all samples and save as an object called `dds_filtered`

dds_filtered <- dds[(rowSums(counts(dds))>0),] #you can subset using the [,] function, you can put commands in there to subset. rowSums counts across the row. By removing 0s we know that there are no genes that are expressed in e.g. group A or B - therefore we would know there is no difference between the two samples. Alternatively, we can also add a threshold to cut of e.g. 1 count per million.

# check dims of dds_filtered  

  str(dds_filtered)
  str(dds)
  
```

# Perform exploratory analysis on dds_filtered object 

1. transform dds_filtered object using `rlog()` functions
    - name your new variables `rlog_transformed_dds` 
    - we want to set `blind = TRUE` as because we are doing a QC step we do not want the function to take into account the experimental design, we want it to be an unbiased look at how the samples cluster
    - check the class of the rlog_transformed_dds objects 
    - you can look at the tranformed counts table using `assay()` 

## NOTE we only use this tranformed data for QC not for downstream analysis or differential expression itself!!


```{r}
#transform data using `rlog()` functions
rlog_transformed_dds <- rlog(dds_filtered, blind=TRUE)

# Notes on `blind` parameter: 
  # blind = TRUE is default
  # blind = TRUE is not the appropriate choice if you expect many or majority of genes to have large differences in counts explainable by   experimental design
  # Mike Love recommends blind = FALSE for visualisation, clustering and machine learning but blind = TRUE for QC
  # Experimental design is only used for calculating the gene-wise dispersion estimates in order to fit a trend line through the dispersions over the mean

  
#check the class of these variable objects 
str(rlog_transformed_dds)
class(rlog_transformed_dds)
  
#get and look at the tranformed counts table using `assay()`
assay(rlog_transformed_dds) #you don't use assays(name)$count as the rlog isn't stored in the count data, it is just added at the end as an unnamed variable.

```

# Optional: If you have time do the tranformation again but using the vst() function 

- As a guide 
    - rlog - more robust than VST when size factors vary widely (10x)
    - VST - less sensitive to high count outliers than rlog, quicker - recommended most often 

```{r}
# get vst transformed data using vst() 
vst_transformed_data <- vst(dds_filtered, blind=TRUE)
assay(vst_transformed_data)

```


# Now plot heatmap of sample-sample correlations using rlog transformed count data and cor()


```{r}
# get correlation matrix of the rlog_transformed counts matrix to get correlations between samples 

cor_matrix_pearson <- cor(assay(rlog_transformed_dds, use = "everything", method = "pearson")) #default = pearson

#NB everything is very highlight correlated

# use pheatmap to plot the correlation matrix 
pheatmap(cor_matrix_pearson) #NB everything is very highlight correlated, make sure you look at the scale, even though it looks as thought there are relationships from the heatmap there is only 0.001 difference in pearson R.
```

# try ploting a heatmap of sample distances 

```{r}
# get sample distances -> Annotated what this code is doing?
sampleDists <- dist(t(assay(rlog_transformed_dds)))  #t() means transpose, dist() is distance
sample_distances_matrix <- as.matrix(sampleDists) 

pretty_blue_colors <- colorRampPalette(rev(brewer.pal(9, "Blues")) )(255)

pheatmap(sample_distances_matrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists, 
         color = pretty_blue_colors)
```

What do you notice here!? what does it tell us about our data? 

Answer: 
There are two sets of duplicats 82 and 88, and 83 and 89, you can see this because they are identical in the above plot.


## You can subset a dds object and interact with it like you would a DataFrame

```{r}
# remove the two suspect columns using [,] notation - remove suspicious samples 
rlog_transformed_dds_subset <- rlog_transformed_dds[,1:6] #can also do [,-(7:8)]

# lets replot the correlation heatmap with the subsetted data - copy code and adjust from above
sampleDists_subset <- dist(t(assay(rlog_transformed_dds_subset)))  #t() means transpose, dist() is distance
sample_distances_matrix <- as.matrix(sampleDists_subset) 

pretty_blue_colors <- colorRampPalette(rev(brewer.pal(9, "Blues")) )(255)

pheatmap(sample_distances_matrix,
         clustering_distance_rows=sampleDists_subset,
         clustering_distance_cols=sampleDists_subset, 
         color = pretty_blue_colors)

```

What do you notice now about the data and the clustering?
They have separated into two separate cell types

# Now lets use the plotPCA function from DESeq2 to plot PCA of rlog_transformed_data_subset

```{r, fig.asp=1}
# plot pca of rlog_transformed_dds using plotPCA()

#first look at the help page for plotPCA - this shows that we need to list the intgroups - i.e. the interesting groups
colData(rlog_transformed_dds_subset) #these are the things we can choose

plotPCA(rlog_transformed_dds_subset, intgroup = c("cell_type", "replicate")) + coord_cartesian() # can use + coord_cartesian() as it is squished flat for some reason, shouldn't normally need it.

# what does the intgroup parameter do? have a go at changing it...

plotPCA(rlog_transformed_dds_subset, intgroup = c("cell_type")) + coord_cartesian()

plotPCA(rlog_transformed_dds_subset, intgroup = c("replicate"))

# how many variable genes is it using for pca? 
#500 as this is the degault

# plot the pca for all genes 
plotPCA(rlog_transformed_dds_subset, intgroup = c("cell_type"), ntop = nrow(rlog_transformed_dds_subset)) + coord_cartesian()
```


What does it tell us about our sample?

Answer: 1 replicate different for CD4


------------- This is the end of our Exploratory data analysis -----------------

We now go back to our original dds object with the raw counts and can update it with anything we need to from the exploratory analysis 
    e.g. - update design with variables in sample info table we want to regress out 
         - remove samples (poor quaility or those with name switches etc)
         
# Revisiting our inital data 

What do we need to do to our data??? 

## Update dds_filtered object if needed - call it dds_filtered_updated

```{r}

# do we need to remove/edit samples? 
dds_filtered_updated <- dds_filtered[,1:6] #this drops the duplicate values we identified.

# do we want to update the design? Is this nessecary in this case?
# e.g. design(dds) <- ~replicate + condition
design(dds_filtered_updated) #all the samples are the same genotype so we just need cell type.

design(dds_filtered_updated) <- ~cell_type

```


# Now we can start our actual DE!!!! yay 

Run deseq2 modelling using one function (DESeq())

```{r}

# look at dds slots then run deseq2 function 
# what has happened to your dds slots?

# run deseq2 modelling on the updated dds object in one function using DEseq() 
dds_filtered_updated <- DESeq(dds_filtered_updated)

# then extract the results
results_obj <- results(dds_filtered_updated)

# sumarise results 
summary(results_obj)

```

Have a look at the outputs 

# Now we will run the analysis step by step to see whats going on 

## 1. Estimate size factors using estimateSizeFactors()

```{r}
# Calculate the size factors for each sample using estimateSizeFactors()
dds_filtered_updated <- estimateSizeFactors(dds_filtered_updated)

# Access the size factors from the dds_filtered_updated object
sizeFactors(dds_filtered_updated) #can see if any of them are very different

# Do you see any systematic differences in size factors between groups?
colData(dds_filtered_updated) #size factor centralises the factors to 1, you are just looking for a very large value compared to the others.
```

# Optional: Plot size factors as barchart using ggplot
```{r}
size_factors_df <- data.frame(colData(dds_filtered_updated)) %>%
  rownames_to_column(var = c("sample_id"))

size_factors_df

ggplot(data = size_factors_df, aes(x=sample_id, y = sizeFactor, fill = cell_type)) + geom_col()
```

## Get counts table normalised using size factors

```{r}
# get counts table normalised using size factors - call it normalized_counts

normalised_counts <- counts(dds_filtered_updated, normalized = TRUE) #counts is a specific function in the DESeq package that has extra things.

# check the class of the output
class(normalised_counts)
# check the class of the output

```

# 3. Estimate dispersions for each gene

```{r}
# Obtain dispersion estimates for each gene using estimateDispersions() 
estdisp_dds <- estimateDispersions(dds_filtered_updated)
plotDispEsts(dds_filtered_updated) # this is a plotting function on estimateDispersion() help page
# Plot the per-gene dispersion estimates (DESeq2 has a helper function for this)

```


## 4. Perform Wald test

```{r}
# Tests for significance of coefficients in a negative binomial GLM, using previously calculated size factors and dispersion estimates and WaldTest
wald_test <- nbinomWaldTest(dds_filtered_updated)

```


# So we have the actual steps that we've gone through to get to DE in one place generate the dds object again and use the DESeq() function to perform steps 1-3 in one go

```{r}
# generate dds object 
dds2 <- DESeqDataSetFromMatrix(countData = counts_table,
                              colData = sample_info,
                              design = ~1) #~1 this is a placeholder, you can change the design later
# filter rows with 0 counts 
dds_filtered2 <- dds[(rowSums(counts(dds))>0),]

# remove unwanted samples
dds_filtered2 <- dds_filtered2[,1:6]

# make DESeqDataSet (dds) object 
design(dds_filtered2) <- ~cell_type

# run DESeq() 
dds_filtered2 <- DESeq(dds_filtered2)
```

# Get the results from your dds object - call it `results_obj`

```{r}
# What do the results() and summary() do, what arguments can you specify? 
results_obj2 <- results(dds_filtered2)
results_obj2 # this says the log fold change cell type (what we told it to compare) CD8 (test) versus CD4 (control, this is why the factor we set at the start is important) You run DESeq once and then you run results multiple times and you can add extra specifications and tests with results function.

summary(results_obj2)
```

# Optional - if we have time: Access the coefficients of the Negative Binomnial GLM – see vingette("DESeq2")
```{r}

```

# Plot normalised counts for ENSMUSG00000041779
 
 checkout the 'plotCounts()` function
 
```{r}
# Plot normalised counts for gene ENSMUSG00000041779
plotCounts(dds_filtered2, gene = "ENSMUSG00000041779", intgroup = "cell_type", normalized = TRUE)
```

# Access the results table for the comparison between CD8 and CD4 T cells . Youu can also try to CD4 vs CD8 (i.e. switch the reference) - you could also try this for replicate 1 vs replicate 3 as a control 

checkout `resultsNames()`  & `results()`

```{r}
?resultsNames()

resultsNames(dds_filtered2)

results_obj_cd4v8 <- results(dds_filtered2, contrast = c("cell_type", "CD4", "CD8")) #by using the contrast you can select the comparisons that you want to make. This is the original way we have before
results(dds_filtered2, contrast = c("cell_type", "CD8", "CD4")) #you can then switch CD4 and CD8, so CD8 is the control. You don't have to go all the way back. You need to set "cell type" in your design. you can have multiple things e.g. replicate and cell type - but you need to specific replicate in the cell design. DESeq calculates everything, and then the results function extracts the information that has already been calculated.
```

Table should contain log2 fold changes and p-values for all genes

# Plot a histogram of the raw and BH-adjusted p-values using ggplot – do they look as expected? for CD8 vs CD4

```{r}
results_obj_cd4v8 <- as.data.frame(results_obj_cd4v8) 

ggplot(results_obj_cd4v8) +
  geom_histogram(aes(x=pvalue), fill = "Blue", alpha = 0.5) +
  geom_histogram(aes(x=padj), fill = "Red", alpha = 0.5) #alpha here is the ggplot function for transparency
```


# Generate an MA plot of the log2 FC values for all genes using DESeq2 plotMA()

```{r}
plotMA(dds_filtered2, alpha = 0.1) #alpha here is the significance level.

plotMA(results_objMA, alpha = 0.1) #NB to run it on the results it has to be the DESeq results object rather than the dataframe we converted.

results_objMA <- results(dds_filtered2, contrast = c("cell_type", "CD4", "CD8"))

```

# Shrink the log2 FC values using the normal, apeglm and ashr methods -> see the vignette – which shall we choose? 

```{r}

# call your results table  called shrunkenLFC_dds_results_obj
shrunkenLFC_dds_results_obj <- lfcShrink(dds_filtered2, coef= "cell_type_CD8_vs_CD4", type="apeglm") #the coef is what we got as an output from a command earlier.

plotMA(shrunkenLFC_dds_results_obj)
```


# Generate a results table (one shrinkage method) containing gene symbols and (if time) entrez ids 

```{r}

# convert shrunkenLFC_dds_results_obj to a dataframe and call it shrunkenLFC_dds_results_df
shrunkenLFC_dds_results_df <- as.data.frame(shrunkenLFC_dds_results_obj)

# Use the org.Mm.eg.db package to get gene symbols for the ensembl gene ids (hint:you did this on bioconductor day) 

# How many Ensembl IDs are not assigned an gene symbol and how many gene symbols are duplicated?



keytypes(org.Mm.eg.db)      #to remind us of what keytypes are in this particular db

# use mapIds() to get 1:1 mapping for SYMBOL, ENTREZ and MGI

# add mapId vectors to  shrunkenLFC_dds_results_df
```


# Export all the results to CSV

```{r}
write_csv(shrunkenLFC_dds_results_df, "CD8_vs_CD4_DE_results_lfcshrinkage_allgenes.csv")
```

# Filter res_df and save as CSV

```{r}
#Filter the results table for padj < 0.05 and absolute log2 FC > 1, and write to a CSV file
res_0.05 <- shrunkenLFC_dds_results_df %>% 
    dplyr::filter(padj < 0.05 & abs(log2FoldChange) > 1)

#Get upregulated genes - Filter the results table for padj < 0.05 and log2FoldChange FC > 1, and write to a CSV file
res_0.05_up <- shrunkenLFC_dds_results_df %>% 
    dplyr::filter(padj < 0.05 & log2FoldChange > 1)

#Get doownregulated genes - Filter the results table for padj < 0.05 and log2FoldChange FC < 1, and write to a CSV file
res_0.05_down <- shrunkenLFC_dds_results_df %>% 
    dplyr::filter(padj < 0.05 & log2FoldChange < -1)

write_csv(res_0.05, 
          "CD8_vs_CD4_DKO_DE_results_lfcshrinkage_0.05_log2FC1.csv")

write_csv(res_0.05_up,
          "CD8_vs_CD4_DKO_DE_results_lfcshrinkage_0.05_log2FC1_up.csv")

write_csv(res_0.05_down,
          "CD8_vs_CD4_DKO_DE_results_lfcshrinkage_0.05_log2FC1_down.csv")
```


Well done! That is the basics of exploratory data analysis for bulk RNAseq and differential gene expression - these statistical methods underly the foundations for single cell differential expression analysis, although scRNA-seq has a few extra paramterisations and complications to take into account. 

We will now progress to pathway/geneset analysis to get some biological meaning from our DE gene list....

