---
title: "Template code for R statistics"
author: "Kevin Rue-Albrecht"
date: "03/10/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(readxl)
```

# Demo

## Mean and standard deviation

The base R functions `mean()` and `sd()` compute the mean and standard deviation of a distribution

To demonstrate, let us first generate a vector of random, normally distributed, values.

```{r}
set.seed(1)
x <- rnorm(n = 100, mean = 2, sd = 5)
```

We can then use that vector to demonstrate the functions.

```{r}
mean(x)
```

```{r}
sd(x)
```


**What are optional arguments for those functions?**

> Answer:

**Why do you think the mean and standard deviation are not exactly those that we would expect?**

> Answer:

# Exercise

## Generate and visualise a distribution

- Generate a vector of 1,000 normally distributed values with mean 10 and standard deviation 5.

```{r}
set.seed(1)
y <- rnorm(1000, mean=10, sd=5)


```

- Inspect the output of the `summary()` function for that vector.

```{r}
summary(y)
```

- Compute the mean and standard deviation for those values.

```{r}
mean(y)
sd(y)

```

```{r}
quantile(y, probs = seq(from = 0.1, to = 1, by = 0.1))
#the seq function is used to produces deciles. The sequence function makes a sequence, in this case we tell it to go from 0.1 to 1 by increments of 0.1

```

- Compute the deciles (i.e. 10 evenly spaced quantiles) for those values.

```{r}
df <- data.frame(y)
```

- Visualise the distribution of those values as a histogram.

```{r}
histogram1 <- ggplot(df, aes(y)) +
              geom_histogram(colour="black")
histogram1
```

- Visualise as vertical lines on the histogram: the mean (red solid), median (red dashed), one standard deviation from the mean (blue solid), and one median absolute deviation from the median (blue dashed).

```{r}
summary_df <- df %>% 
              summarise(
                mean = mean(y),
                median = median(y),
                sd = sd(y),
                mad = mad(y)
                )
summary_df

histogram1 <- ggplot(df, aes(y)) +
              geom_histogram(colour = "black") +
              geom_vline(
                xintercept = c(
                  summary_df$mean,
                  summary_df$median,
                  summary_df$mean + summary_df$sd,
                  summary_df$mean - summary_df$sd,
                  summary_df$median + summary_df$mad,
                  summary_df$median - summary_df$mad
                  ),
                colour = c(
                  "red",
                  "blue",
                  "blue",
                  "blue",
                  "blue",
                  "blue"
                  ),
                linetype = c(
                  1,
                  2,
                  1,
                  1,
                  2,
                  2
                )
              )
histogram1



```

- Generate a new vector with _a lot_ more values (e.g., one million).
  Draw again a histogram.
  How does the distribution compare with more data points?

```{r}
set.seed(1)
x <- rnorm(1000000, mean=10, sd=5)
df_x <- data.frame(x)

summary_x <- df_x %>% 
              summarise(
                mean = mean(x),
                median = median(x),
                sd = sd(x),
                mad = mad(x)
                )

histogram_x <- ggplot(df_x, aes(x)) +
              geom_histogram(colour = "black") +
              geom_vline(
                xintercept = c(
                  summary_df$mean,
                  summary_df$median,
                  summary_df$mean + summary_df$sd,
                  summary_df$mean - summary_df$sd,
                  summary_df$median + summary_df$mad,
                  summary_df$median - summary_df$mad
                  ),
                colour = c(
                  "red",
                  "blue",
                  "blue",
                  "blue",
                  "blue",
                  "blue"
                  ),
                linetype = c(
                  1,
                  2,
                  1,
                  1,
                  2,
                  2
                )
              )
histogram_x
```

## Query distributions and probabilities

For the standard normal distribution ${\mathcal {N}}(\mu=0 ,\sigma ^{2}=1)$:

- Plot the cumulative distribution function in the range $[-5, 5]$ in 0.1 increment.

```{r}
df_cumul <- data.frame(x = seq(from = -5, to = 5, by = 0.1))

plot_cumul <- ggplot(df_cumul, aes(x = x, y = pnorm(x))) +
              geom_point()

#pnorm is the cumulative distribution

plot_cumul

```

- Plot the inverse cumulative distribution function for quantiles in 0.01 increment.

```{r}

plot_qnorm <- ggplot(df_cumul, aes(x = x, y = qnorm(x))) +
              geom_point()

#qnorm is the inverse cumulative distribution function
# Q norm wants a probability and it give an number out put so it only runds from 0 to 1. All the values <0 and >1 will give NaN.

plot_qnorm

```

- Plot the density function in the range $[-5, 5]$ in 0.1 increment.

```{r}

plot_dnorm <- ggplot(df_cumul, aes(x = x, y = dnorm(x))) +
              geom_point()

#dnorm is the density function 

plot_dnorm

```

- What is the probability of observing a value greater than 2?

```{r}
1 - pnorm(2) #this uses the default settings in R, with your data set you would give the mean and sd (look at the help page).
```

- What is the probability of observing a value between -2 and 2?

```{r}
pnorm (2) - pnorm(-2)

#this calculates the probability of the bit in the middle of a normal distribution scale between these two values.
```

- What is the probability of observing a value more extreme than -2 or 2?

```{r}
1- (pnorm (2) - pnorm(-2))

#this calculates the probability of the two tails when the bit in the middle of the normal distribution is removed.
```


### Run a Wilcoxon test to compare the vectors 1:10 and 5:14
```{r}

a <- c(1:10)
b <- c(5:14)

wilcox.test(c(1:10),c(5:14), paired = FALSE)
wilcox.test(a,b, paired = FALSE)

#both of these produce the same value, the warning message occurs as the value increases by 1 each time there is no randomness in this data.
```

### T test

```{r}
gene_df <- read_csv("/project/obds/shared/resources/3_r_stats_genomics/statistics/data/gene_exprs.csv")

head(gene_df)
```
Run a t-test to compare the gene expression values between the two groups.

```{r}
#first check you only have two groups
unique(gene_df$group)

#then run the t test
t.test(gene_exprs ~ group, data = gene_df)
```

### You can also use base R
```{r}
t.test(
  x = gene_df$gene_exprs[gene_df$group == "groupA"],
  y = gene_df$gene_exprs[gene_df$group == "groupB"]
)
```
### lapply returns a vector (has to be scalars) whereas loop returns an object.
```{r}
library(readxl)
log_count_matrix <- read_excel("/project/obds/shared/resources/3_r_stats_genomics/statistics/data/GSE111543.xlsx")

#if you open without specifying, it will open the first sheet.
# can shee what sheets there are by using: excel_sheets("/project/obds/shared/resources/3_r_stats_genomics/statistics/data/GSE111543.xlsx")

#apply only work properly on a matrix, if anything is a character everything will be a character. So you need to convert to a dataframe, then move the gene name to row names, then convert to a matrix.

log_count_matrix <- as.data.frame(log_count_matrix)

log_count_matrix <- column_to_rownames(log_count_matrix, "gene")

log_count_matrix <- as.matrix(log_count_matrix)

head(log_count_matrix)

```

Look at the sheets
```{r}
excel_sheets("/project/obds/shared/resources/3_r_stats_genomics/statistics/data/GSE111543.xlsx")
```

```{r}
cell_info <- read_excel("/project/obds/shared/resources/3_r_stats_genomics/statistics/data/GSE111543.xlsx", sheet = "cell_info")
```

```{r}
unique(cell_info$Infection)
```

```{r}
infected <- cell_info %>% 
  filter(Infection == "STM-LT2") %>%
  pull(Sample) #use pull rather than select so that you get it as a vector 

notinfected <- cell_info %>% 
  filter(Infection == "Mock") %>%
  pull(Sample)

notinfected
```
### take the first row first to check
```{r}
exprs_infected <- log_count_matrix[1, infected] 
#you can extract the columns by their name as you have already substed them in the 'infected' object

exprs_notinfected <- log_count_matrix[1, notinfected] 
#you can extract the columns by their name as you have already substed them in 'infected'

exprs_infected
exprs_notinfected
```

```{r}
row_1_test <-  t.test(x = exprs_infected, y = exprs_notinfected)
row_1_test
```
### the results of all tests are just a list displayed in a fancy way. You can extract bits of a test by applying str(testname) on the test. This then gives you all the $variable for the test. You can then extract these using testname$variable. We can use this to extract the p value.

```{r}
str(row_1_test)

row_1_test$p.value
```

### 3. Use the function vapply to test every gene set in the list and collect a vector of p-values.

```{r}
t_test_gene <- function(index) {
                   #index is like x, it is the value of the row name.
  exprs_infected <- log_count_matrix[index, infected] 
  exprs_notinfected <- log_count_matrix[index, notinfected]
  index_test <- t.test(x = exprs_infected, y = exprs_notinfected) #it then applies a t test for each of those genes
  index_test$p.value
  
  return(index_test$p.value) #return is the output you want it to produce
}

#example gene and it outputs the p value
t_test_gene(index = "ENSG00000237973")

```


```{r}
gene_rownames <- rownames(log_count_matrix)

gene_rownames

pvaluesgenes <- vapply(X = gene_rownames,
                  FUN = t_test_gene,
                  FUN.VALUE = numeric(1)
                  ) 
#Fun.value = we want a numeric value of length 1, can also do numeric(0), which means you just expect a numeric value.
#vapply is more efficient than sapply as it is designed to be applied over a vector1.

head(pvaluesgenes)
```
















# Demo

## Empirical Cumulative Distribution Function

```{r}
ecdf_iris_sepal_length <- ecdf(iris$Sepal.Length)
ecdf_iris_sepal_length
```

```{r}
ggplot(iris, aes(Sepal.Length)) +
  geom_histogram(color = "black", fill = "grey") +
  cowplot::theme_cowplot()
```

```{r}
plot(ecdf_iris_sepal_length, cex = 0.5)
```

# Demo

## ecdf - Knots

```{r}
knots(ecdf_iris_sepal_length)
```

```{r}
sort(unique(iris$Sepal.Length))
```

# Demo

## ecdf - Quantiles

```{r}
quantile(ecdf_iris_sepal_length, c(0, 0.25, 0.5, 0.75, 1))
```

```{r}
quantile(iris$Sepal.Length, c(0, 0.25, 0.5, 0.75, 1))
```

# Demo

## <i class="fab fa-r-project"></i> Functions for Statistical Testing

```{r}
?pairwise.t.test
```

```{r}
help(pairwise.t.test)
```

# Demo

## Parametric t-test

Two normal distributions.

```{r}
set.seed(1)
x <- rnorm(n = 1000, mean = 0, sd = 1)
y <- rnorm(n = 1000, mean = 1, sd = 1)
```

```{r}
test_data <- bind_rows(
  tibble(group = "x", value = x),
  tibble(group = "y", value = y)
)
ggplot(test_data, aes(value)) +
  geom_histogram(fill = "grey", color = "black") +
  facet_wrap(~group, ncol = 1) +
  cowplot::theme_cowplot()
```

Unpaired t-test.

```{r}
t.test(value ~ group, test_data)
```

Compare with

```{r}
t.test(x, y)
```

```{r}
t.test(y, x)
```

# Demo

## Non-parametric wilcoxon test

Two uniform distributions

```{r}
set.seed(1)
x <- runif(n = 1000, min = 1, max = 11)
y <- runif(n = 1000, min = 3, max = 13)
```

```{r}
test_data <- bind_rows(
  tibble(group = "x", value = x),
  tibble(group = "y", value = y)
)
gg <- ggplot(test_data, aes(value)) +
  facet_wrap(~group, ncol = 1) +
  geom_histogram(fill = "grey", color = "black") +
  cowplot::theme_cowplot()
gg
```
]

Mann-Whitney U test

```{r}
wilcox.test(value ~ group, test_data)
```

Directed hypothesis

```{r}
wilcox.test(value ~ group, test_data, alternative = "less")
```

# Demo

## Paired test

For each sample, the two measurements are related to one another; e.g. patients measured before and after a treatment.

```{r}
set.seed(1)
n_sample <- 10
x <- runif(n = n_sample, min = 10, max = 20)
y <- x + 2 + rnorm(n = n_sample, mean = 0, sd = 1)
```

```{r}
test_data <- tibble(
  sample = paste("sample", seq_len(n_sample)),
  x = x,
  y = y
) %>% pivot_longer(cols = c(x, y), names_to = "variable")
```

```{r}
ggplot(test_data, aes(variable, value)) +
  geom_line(aes(group = sample), size = 0.1) +
  geom_point() +
  cowplot::theme_cowplot()
```

```{r}
t.test(x, y, paired = TRUE)
```

# Demo

## Analysis of Variance (ANOVA)

```{r}
set.seed(1)
n_sample <- 1000
x1 <- rnorm(n = n_sample, mean = 10, sd = 2)
x2 <- x1 + 5 + rnorm(n = n_sample, mean = 0, sd = 1)
x3 <- x2 + 0 + rnorm(n = n_sample, mean = 0, sd = 0.5)
test_data <- bind_rows(
  tibble(group = "x1", value = x1),
  tibble(group = "x2", value = x2),
  tibble(group = "x3", value = x3)
)
```

```{r}
test_data <- bind_rows(
  tibble(group = "x1", value = x1),
  tibble(group = "x2", value = x2),
  tibble(group = "x3", value = x3)
)
gg <- ggplot(test_data, aes(value)) +
  facet_wrap(~group, ncol = 1) +
  geom_histogram(fill = "grey", color = "black") +
  cowplot::theme_cowplot()
gg
```

```{r}
out <- aov(value ~ group, test_data)
out
```

```{r}
summary(out)
```

# Demo

## Linear models

```{r}
set.seed(1)
test_data <- tibble(
  x = rnorm(n = 50, mean = 0, sd = 1),
  y = 10 + 2.5 * x + rnorm(n = 50, mean = 0, sd = 0.5))
```

```{r}
ggplot(test_data, aes(x = x, y = y)) +
  geom_point() +
  stat_smooth(method = "glm", se = FALSE) +
  cowplot::theme_cowplot()
```

```{r}
lm(y ~ x, test_data)
```

# Demo

## Linear models - summary

```{r}
lm(y ~ x, test_data) %>% summary()
```

# Demo

## Fisher's Exact Test

```{r}
x_table <- matrix(data = c(12, 4, 3, 23),
  nrow = 2,
  dimnames = list(
    c("DE","Not DE"),
    c("in_pathway", "not_pathway")
  ))
knitr::kable(x_table)
```

```{r}
fisher.test(x_table)
```

# Demo

## Beware of interpreting inadequate tests!

Two uniform distributions.

```{r}
set.seed(1)
n_size <- 10E3
x <- runif(n = n_size, min = 1, max = 11)
y <- runif(n = n_size, min = 3, max = 13)
```

```{r}
test_data <- bind_rows(
  tibble(group = "x", value = x),
  tibble(group = "y", value = y)
)
gg <- ggplot(test_data, aes(value)) +
  facet_wrap(~group, ncol = 1) +
  geom_histogram(fill = "grey", color = "black") +
  cowplot::theme_cowplot()
gg
```

Nothing prevents users from running a parametric test.

```{r}
t.test(value ~ group, test_data)
```

# Demo

## Knowledge assumptions - Central tendency

```{r, message=FALSE, warning=FALSE}
x_mean <- 0
x_sd <- 20
data_table <- tibble(x = as.integer(rnorm(n = 10E3, mean = x_mean, sd = x_sd)))
summary_table <- bind_rows(
  tibble(Value = "mean", value = mean(data_table$x)),
  tibble(Value = "median", value = median(data_table$x)),
  tibble(Value = "mode", value = as.integer(names(which.max(table(data_table$x)))))
)
data_table %>% 
  ggplot() +
  geom_histogram(aes(x = x), color = "black", fill = "grey") +
  geom_vline(aes(xintercept = value, color = Value), summary_table, size = 2, alpha = 0.3) +
  cowplot::theme_cowplot()
```

# Demo

## Knowledge assumptions - Normality

Normal distribution

```{r}
x <- rnorm(n = 5000, mean = 0, sd = 1)
```

```{r}
ggplot(tibble(x=x)) +
  geom_histogram(aes(x), fill = "grey", color = "black", bins = 20) +
  cowplot::theme_cowplot()
```

```{r}
shapiro.test(x)
```

Log-normal distribution

```{r}
x <- 2^rnorm(n = 5000, mean = 0, sd = 1)
```

```{r}
ggplot(tibble(x=x)) +
  geom_histogram(aes(x), fill = "grey", color = "black", bins = 20) +
  cowplot::theme_cowplot()
```

```{r}
shapiro.test(x)
```

# Demo

## Knowledge assumptions - Normality

Normal distribution

```{r}
x <- rnorm(n = 5000, mean = 5, sd = 3)
qqnorm(x)
```

Log-normal distribution

```{r}
x <- 2^rnorm(n = 5000, mean = 0, sd = 1)
qqnorm(x)
```

# Demo

## Multiple-testing correction

For 1,000 iterations, generate two vectors of normally distributed values with mean 0 and standard deviation 1, and run a t-test to compare the two.

```{r}
set.seed(1)
n_tests <- 1000
compute_p_value <- function(dummy) {
  x <- rnorm(n = 100, mean = 0, sd = 1)
  y <- rnorm(n = 100, mean = 0, sd = 1)
  out <- t.test(x, y)
  out$p.value
}
result_table <- tibble(
  pvalue = vapply(X = seq_len(n_tests), FUN = compute_p_value, FUN.VALUE = numeric(1)),
  BH = p.adjust(p = pvalue, method = "BH"),
  bonferroni = p.adjust(p = pvalue, method = "bonferroni")
)
```

Illustrate one iteration.

```{r, include=TRUE, echo=FALSE, fig.height=3}
data_table <- tibble(
  x = rnorm(n = 100, mean = 0, sd = 1),
  y = rnorm(n = 100, mean = 0, sd = 1)
) %>% pivot_longer(cols = c(x, y))
ggplot(data_table) +
  geom_boxplot(aes(name, value)) +
  geom_jitter(aes(name, value), width = 0.1)
```

Plot the distribution of raw p-values.

```{r, include=TRUE, echo=FALSE, fig.height=3}
ggplot(result_table) +
  geom_histogram(aes(pvalue), fill = "grey", color = "black", breaks = seq(0, 1, 0.05)) +
  scale_x_continuous(limits = c(0, 1)) +
  labs(title = "Raw p-value")
```

Print a short report message.

```{r, include=TRUE, echo=FALSE, results='asis'}
cat(sprintf("There are %i out of %i raw p-values smaller than 0.05", sum(result_table$pvalue < 0.05), n_tests))
```

Plot corrected p-values using BH correction.

```{r, include=TRUE, echo=FALSE, fig.height=3}
ggplot(result_table) +
  geom_histogram(aes(BH), fill = "grey", color = "black", bins = 20) +
  coord_cartesian(xlim = c(0, 1)) +
  labs(title = "BH correction")
```

Print a short report message.

```{r, include=TRUE, echo=FALSE, results='asis'}
cat(sprintf("There are %i BH-corrected p-values smaller than 0.05", sum(result_table$BH < 0.05)))
```

Plot corrected p-values using Bonferroni correction.

```{r, include=TRUE, echo=FALSE, fig.height=3}
ggplot(result_table) +
  geom_histogram(aes(bonferroni), fill = "grey", color = "black", bins = 20) +
  coord_cartesian(xlim = c(0, 1)) +
  labs(title = "bonferroni correction")
```

Print a short report message.

```{r, include=TRUE, echo=FALSE, results='asis'}
cat(sprintf("There are %i bonferonni corrected p-values smaller than 0.05", sum(result_table$bonferroni < 0.05)))
```

# Exercise

## Testing & Multiple testing correction

Given an Excel file that contains a matrix of log-normalised counts (`logcounts`) and experimental metadata (`cell_info`),
test each gene (i.e., row) in the matrix for differential expression between the two experimental groups.
Start by importing the `logcounts` table and converting it to a matrix.

```{r}
logcounts_matrix <- read_excel(   )


dim(logcounts_matrix)
```

```{r}
cell_info <- read_excel(   )
# all(rownames(cell_info$Sample) == colnames(logcounts_matrix))
head(cell_info)
```

### Approach

1. Write the code to test a single gene and access the p-value.

```{r}
gene_index <- 1




out <- t.test(   )
out$p.value
```

2. Write a function that generalises the code to test any one gene and return the p-value.

```{r}
t_test_row <- function(index,    ) {






}
t_test_row(index = 1,    )
```

3. Use the function `vapply` to test every row in the matrix and collect a vector of p-values.

```{r}
t_test_pvalues <- vapply(
  X =    ,
  FUN =     ,
  FUN.VALUE =     ,

)

head(t_test_pvalues)
```

### Bonus points

- Visualise a histogram of the p-values.

```{r}

```

- Correct p-values for multiple testing.
  How many genes remain before and after multiple testing?

```{r}
result_bh <- 

```

- Use `gene_info` to get the gene name for the gene identifier with the smallest p-value.

```{r}
gene_table <- read_excel(    )


```

# Exercise

## Over-representation analysis (ORA)

Given the list of genes (Ensembl gene identifiers) that your identified as differentially expressed in the previous exercise,
and a list of gene sets (`go_db`),
test each gene set for over-representation of differentially expressed genes.
Start by importing the Gene Ontology table and converting it to a list.

```{r}
go_table <- read_excel(    )
go_list <- split(go_table$ensembl_gene_id, go_table$go_id)
go_list[1]
```

### Approach

1. Write the code to test a single gene set and access the p-value.

```{r}
geneset <- go_list[[1]]









out <- fisher.test(   )
out$p.value
```

2. Write a function that generalises the code to test any one gene set and return the p-value.

```{r}
fisher_test_pathway <- function(index,    ) {










}
fisher_test_pathway(index = 1,    )
```

3. Use the function `vapply` to test every gene set in the list and collect a vector of p-values.

```{r}
fisher_test_pvalues <- vapply(
  X =    ,
  FUN =    ,
  FUN.VALUE =    ,
  
  
)  

head(fisher_test_pvalues)
```

### Bonus points

- Visualise a bar plot of the p-values.

```{r}

```

- Correct p-values for multiple testing.
  How many gene sets remain before and after multiple testing?

```{r}
fisher_test_bh <- 

```

- Use `go_info` to annotate each GO gene set with its corrected p-value,
  and arrange the table by increasing p-value.

```{r}
go_info <- read_excel(   )



```
